diff --git a/drivers/tty/vt/selection.c b/drivers/tty/vt/selection.c
index 8d7ab8809..abbaa10a7 100644
--- a/drivers/tty/vt/selection.c
+++ b/drivers/tty/vt/selection.c
@@ -348,17 +348,6 @@ int set_selection(const struct tiocl_selection __user *v, struct tty_struct *tty
 	return ret;
 }
 
-int set_selection(const struct tiocl_selection __user *v,
-	struct tty_struct *tty)
-{
-	int ret;
-
-	console_lock();
-	ret = __set_selection(v, tty);
-	console_unlock();
-
-	return ret;
-}
 
 /* Insert the contents of the selection buffer into the
  * queue of the tty associated with the current console.
