diff --git a/drivers/power/supply/mediatek/battery/mtk_battery.c b/drivers/power/supply/mediatek/battery/mtk_battery.c
index 8ea3a946b..757cfee83 100644
--- a/drivers/power/supply/mediatek/battery/mtk_battery.c
+++ b/drivers/power/supply/mediatek/battery/mtk_battery.c
@@ -72,7 +72,6 @@
 #include <mtk_gauge_time_service.h>
 #include <mt-plat/upmu_common.h>
 #include <pmic_lbat_service.h>
-#include <mt-plat/mtk_battery.h>
 /* ============================================================ */
 /* define */
 /* ============================================================ */
@@ -118,7 +117,6 @@ static int battery_out_data[1] = { 0 };
 static int fcc = 6000000;
 static int thermal_input_current = 3000000;
 static bool g_ADC_Cali;
-static struct wakeup_source status_wakelock;
 
 struct power_supply *max_verify_psy;
 
@@ -128,26 +126,27 @@ static enum power_supply_property battery_props[] = {
 	POWER_SUPPLY_PROP_PRESENT,
 	POWER_SUPPLY_PROP_TECHNOLOGY,
 	POWER_SUPPLY_PROP_CYCLE_COUNT,
-	POWER_SUPPLY_PROP_CAPACITY,
+	POWER_SUPPLY_PROP_VOLTAGE_NOW,
 	POWER_SUPPLY_PROP_CURRENT_NOW,
 	POWER_SUPPLY_PROP_CURRENT_AVG,
-	POWER_SUPPLY_PROP_VOLTAGE_NOW,
+	POWER_SUPPLY_PROP_CHARGE_FULL_DESIGN,
 	POWER_SUPPLY_PROP_CHARGE_FULL,
 	POWER_SUPPLY_PROP_CHARGE_COUNTER,
-	POWER_SUPPLY_PROP_TEMP,
-	POWER_SUPPLY_PROP_SET_TEMP_ENABLE,
-        POWER_SUPPLY_PROP_SET_TEMP_NUM,
-	POWER_SUPPLY_PROP_INPUT_SUSPEND,
-	POWER_SUPPLY_PROP_CAPACITY_LEVEL,
-	POWER_SUPPLY_PROP_TIME_TO_FULL_NOW,
-	POWER_SUPPLY_PROP_CHARGE_FULL_DESIGN,
 	POWER_SUPPLY_PROP_CHARGE_CONTROL_LIMIT,
 	POWER_SUPPLY_PROP_CHARGE_CONTROL_LIMIT_MAX,
 	POWER_SUPPLY_PROP_FAST_CHARGE_CURRENT,
 	POWER_SUPPLY_PROP_THERMAL_INPUT_CURRENT,
+	POWER_SUPPLY_PROP_CAPACITY,
+	POWER_SUPPLY_PROP_CAPACITY_LEVEL,
+	POWER_SUPPLY_PROP_TEMP,
+	POWER_SUPPLY_PROP_TIME_TO_FULL_NOW,
 	POWER_SUPPLY_PROP_CHARGING_ENABLED,
 	POWER_SUPPLY_PROP_TYPEC_MODE,
+	POWER_SUPPLY_PROP_INPUT_SUSPEND,
+	POWER_SUPPLY_PROP_SET_TEMP_ENABLE,
+	POWER_SUPPLY_PROP_SET_TEMP_NUM,
 	POWER_SUPPLY_PROP_SHUTDOWN_DELAY,
+	POWER_SUPPLY_PROP_REAL_SOC,
 };
 
 /* weak function */
@@ -287,14 +286,50 @@ bool is_battery_init_done(void)
 	return gm.is_probe_done;
 }
 
+struct tag_bootmode {
+	u32 size;
+	u32 tag;
+	u32 bootmode;
+	u32 boottype;
+};
+
+int battery_get_boot_mode(void)
+{
+	struct device *dev = NULL;
+	struct device_node *boot_node = NULL;
+	struct tag_bootmode *tag = NULL;
+	int boot_mode = 11;//UNKNOWN_BOOT
+
+	dev = gm.gdev->dev.parent;
+	if (dev != NULL) {
+		boot_node = of_parse_phandle(dev->of_node, "bootmode", 0);
+		if (!boot_node) {
+			bm_err("%s: failed to get boot mode phandle\n", __func__);
+		} else {
+			tag = (struct tag_bootmode *)of_get_property(boot_node,
+								"atag,boot", NULL);
+	bm_err("[CHECK!]%s: tag:%d\n", __func__, tag);
+			if (!tag)
+				bm_err("%s: failed to get atag,boot\n", __func__);
+			else {
+				boot_mode = tag->bootmode;
+				gm.boot_mode = tag->bootmode;
+				bm_err("[CHECK!] %s: tag->bootmode:%d, boot_mode:%d, gm.boot_mode:%d\n", __func__, tag->bootmode, boot_mode, gm.boot_mode);
+			}
+		}
+	}
+	bm_err("%s: boot mode=%d\n", __func__, boot_mode);
+	return boot_mode;
+}
+
 bool is_recovery_mode(void)
 {
-	int boot_mode = get_boot_mode();
+	int boot_mode = battery_get_boot_mode();
 
 	if (is_fg_disabled())
 		return false;
 
-	bm_debug("mtk_battery boot mode =%d\n", boot_mode);
+	bm_err("mtk_battery boot mode =%d\n", boot_mode);
 	if (boot_mode == RECOVERY_BOOT) {
 		gm.log_level = BMLOG_DEBUG_LEVEL;
 		fg_cust_data.daemon_log_level = BMLOG_DEBUG_LEVEL;
@@ -453,13 +488,13 @@ static int bms_get_property(struct power_supply *psy,
 		pr_info("gm.battery_id :%d.\n", gm.battery_id);
 		switch (gm.battery_id) {
 		case 0:
-			val->strval = "k7s_nvt";
+			val->strval = "k7sr_nvt";
 			break;
 		case 1:
-			val->strval = "k7s_gy";
+			val->strval = "k7sr_gy";
 			break;
 		case 2:
-			val->strval = "k7s_xwd";
+			val->strval = "k7sr_xwd";
 			break;
 		case 3:
 			val->strval = "unknown";
@@ -514,16 +549,11 @@ static int bms_get_property_maxim(struct power_supply *psy,
 	int rc = 0;
 	union power_supply_propval b_val = {0,};
 	if (max_verify_psy == NULL && suppld_maxim)
-		max_verify_psy = power_supply_get_by_name("batt_verify");
-	if ((psp == POWER_SUPPLY_PROP_AUTHENTIC)
-		|| (psp == POWER_SUPPLY_PROP_ROMID)
-		|| (psp == POWER_SUPPLY_PROP_DS_STATUS)
-		|| (psp == POWER_SUPPLY_PROP_PAGE0_DATA)) {
-		if (max_verify_psy == NULL || (!suppld_maxim)) {
+			max_verify_psy = power_supply_get_by_name("batt_verify");
+	if (max_verify_psy == NULL || (!suppld_maxim)) {
 			pr_err("max_verify_psy is NULL\n");
 			return -ENODATA;
 		}
-	}
 
 	chr_type = mt_get_charger_type();
 
@@ -632,13 +662,13 @@ static int bms_get_property_maxim(struct power_supply *psy,
 		pr_info("gm.battery_id :%d.\n", gm.battery_id);
 		switch (gm.battery_id) {
 		case 0:
-			val->strval = "k7s_nvt_5000mAh";
+			val->strval = "k7sr_nvt_5000mAh";
 			break;
 		case 1:
-			val->strval = "k7s_gy_5000mAh";
+			val->strval = "k7sr_gy_5000mAh";
 			break;
 		case 2:
-			val->strval = "k7s_xwd_5000mAh";
+			val->strval = "k7sr_xwd_5000mAh";
 			break;
 		case 3:
 			val->strval = "unknown";
@@ -831,7 +861,7 @@ void battery_update_psd(struct battery_data *bat_data)
 	bat_data->BAT_batt_temp = battery_get_bat_temperature();
 }
 
-static int battery_get_property(struct power_supply *psy,
+static int battery_psy_get_property(struct power_supply *psy,
 	enum power_supply_property psp,
 	union power_supply_propval *val)
 {
@@ -839,7 +869,10 @@ static int battery_get_property(struct power_supply *psy,
 	int fgcurrent = 0;
 	bool b_ischarging = 0;
 	int input_suspend;
+	struct power_supply	*usb_psy;
 	union power_supply_propval pval = {0};
+	union power_supply_propval real_type = {0};
+	union power_supply_propval pd_active = {0};
 	/* 2020.12.31 longcheer jiangshitian bat FAMMI start */
 	union power_supply_propval chg_flag_unin = {0,};
 	/* 2020.12.31 longcheer jiangshitian bat FAMMI end */
@@ -861,10 +894,36 @@ static int battery_get_property(struct power_supply *psy,
 		break;
 	case POWER_SUPPLY_PROP_STATUS:
 		val->intval = data->BAT_STATUS;
+#if 0
 		input_suspend = charger_manager_is_input_suspend();
 		if (input_suspend || chr_type == CHARGER_UNKNOWN)
 			val->intval = POWER_SUPPLY_STATUS_DISCHARGING;
+#endif
+//ww add charging status start
+		input_suspend = charger_manager_is_input_suspend();
+		usb_psy = power_supply_get_by_name("usb");
+		if(!usb_psy){
+			pr_err("usb_psy is NULL!\n");
+			return 0;
+		}
+		power_supply_get_property(usb_psy,
+		POWER_SUPPLY_PROP_REAL_TYPE, &real_type);
+		power_supply_get_property(usb_psy,
+		POWER_SUPPLY_PROP_PD_ACTIVE, &pd_active);
+		if(val->intval == POWER_SUPPLY_STATUS_FULL){
+			val->intval = POWER_SUPPLY_STATUS_FULL;
+		}else if(input_suspend){
+			val->intval = POWER_SUPPLY_STATUS_DISCHARGING;
+		}else if(((val->intval == POWER_SUPPLY_STATUS_DISCHARGING) ||
+			(val->intval == POWER_SUPPLY_STATUS_NOT_CHARGING)) && (real_type.intval > 0)){
+				val->intval = POWER_SUPPLY_STATUS_CHARGING;
+		}else if(pd_active.intval){
+			val->intval = POWER_SUPPLY_STATUS_CHARGING;
+		}else if(data->BAT_batt_vol < 3300){
+			val->intval = POWER_SUPPLY_STATUS_DISCHARGING;
+		}
 		break;
+//ww add charging status end
 	case POWER_SUPPLY_PROP_HEALTH:
 		val->intval = data->BAT_HEALTH;/* do not change before*/
 		break;
@@ -1008,27 +1067,55 @@ static int battery_get_property(struct power_supply *psy,
 		break;
 	case POWER_SUPPLY_PROP_SHUTDOWN_DELAY:
 		power_supply_get_property(data->psy, POWER_SUPPLY_PROP_CAPACITY, &pval);
-		if (pval.intval <= 1){
-			if (data->BAT_STATUS == POWER_SUPPLY_STATUS_CHARGING){
-				val->intval = 0;	
-			}
-			else if (data->BAT_STATUS != POWER_SUPPLY_STATUS_CHARGING){
-                          	val->intval = 1;
+		gauge_get_current(&fgcurrent);
+		if (pval.intval > 0 && pval.intval <= 1){
+			pr_info("vbat_uv: %d, is_charging: %d, shutdown_delay0: %d, UI_soc:%d",	data->BAT_batt_vol, data->BAT_STATUS, val->intval,pval.intval);
+		/*add for HTH-268226 by changhongjie at 2023/1/9 start*/
+			if (data->BAT_STATUS == POWER_SUPPLY_STATUS_CHARGING && val->intval){
+				if (data->BAT_batt_vol <= 3300) {
+					pr_info("vbat_uv: %d, is_charging: %d, shutdown_delay1: %d, UI_soc:%d",	data->BAT_batt_vol, data->BAT_STATUS, val->intval,pval.intval);
+					kernel_power_off();
+				} else {
+					val->intval = 0;
+				}
+				data->BAT_CAPACITY = 1;
+			} else {
+				if (data->BAT_batt_vol > 3450) {
+					data->BAT_CAPACITY = 1;
+				} else if (data->BAT_batt_vol > 3300) {
+					if (data->BAT_STATUS != POWER_SUPPLY_STATUS_CHARGING){
+						val->intval = 1;
+						pr_info("vbat_uv: %d, is_charging: %d, shutdown_delay2: %d, UI_soc:%d",data->BAT_batt_vol, data->BAT_STATUS, val->intval, pval.intval);
+					}
+					data->BAT_CAPACITY = 1;
+				} else {
+					pr_info("vbat_uv: %d, is_charging: %d, shutdown_delay3: %d, UI_soc:%d",	data->BAT_batt_vol, data->BAT_STATUS, val->intval,pval.intval);
+					kernel_power_off();
+				}
 			}
+		}else if(data->BAT_batt_vol <= 3320 && fgcurrent > 0){
+			pr_info("vbat_uv: %d, is_charging: %d, shutdown_delay4: %d, UI_soc:%d",	data->BAT_batt_vol, data->BAT_STATUS, val->intval,pval.intval);
+			val->intval = 1;
+		}else{
+			val->intval = 0;
 		}
-		else{
-			val->intval = 0;	
+		if(gm.tbat_precise < 50 && data->BAT_CAPACITY >1){
+			pr_info("temp is %d, too low and uisoc is %d, not 0\n",gm.tbat_precise,data->BAT_CAPACITY);
+			val->intval = 0;
 		}
+        /*add for HTH-268226 by changhongjie at 2023/1/9 end*/
+		break;
+	case POWER_SUPPLY_PROP_REAL_SOC:
+		val->intval = battery_get_soc();
 		break;
 	default:
 		ret = -EINVAL;
 		break;
 	}
-
 	return ret;
 }
 
-static int battery_set_property(struct power_supply *psy,
+static int battery_psy_set_property(struct power_supply *psy,
 			enum power_supply_property psp,
 			const union power_supply_propval *val)
 {
@@ -1081,9 +1168,13 @@ static int battery_prop_is_writeable(struct power_supply *psy,
 			enum power_supply_property psp)
 {
 	switch (psp) {
+	case POWER_SUPPLY_PROP_SET_TEMP_ENABLE:
+	case POWER_SUPPLY_PROP_SET_TEMP_NUM:
 	case POWER_SUPPLY_PROP_INPUT_SUSPEND:
-		return 1;
 	case POWER_SUPPLY_PROP_CHARGING_ENABLED:
+	case POWER_SUPPLY_PROP_CHARGE_CONTROL_LIMIT:
+	case POWER_SUPPLY_PROP_FAST_CHARGE_CURRENT:
+	case POWER_SUPPLY_PROP_THERMAL_INPUT_CURRENT:
 		return 1;
 	default:
 		break;
@@ -1099,8 +1190,8 @@ struct battery_data battery_main = {
 		.type = POWER_SUPPLY_TYPE_BATTERY,
 		.properties = battery_props,
 		.num_properties = ARRAY_SIZE(battery_props),
-		.get_property = battery_get_property,
-		.set_property = battery_set_property,
+		.get_property = battery_psy_get_property,
+		.set_property = battery_psy_set_property,
 		.property_is_writeable = battery_prop_is_writeable,
 		},
 
@@ -1230,7 +1321,7 @@ void battery_update(struct battery_data *bat_data)
 
 bool is_kernel_power_off_charging(void)
 {
-	int boot_mode = get_boot_mode();
+	int boot_mode = battery_get_boot_mode();
 
 	if (boot_mode == KERNEL_POWER_OFF_CHARGING_BOOT
 	    || boot_mode == LOW_POWER_OFF_CHARGING_BOOT) {
@@ -1597,12 +1688,12 @@ static void dump_daemon_table(struct seq_file *m)
 		seq_puts(m, "idx: maH, voltage, R1, R2, percentage\n");
 		ptr = &ptable2->fg_profile[j].fg_profile[0];
 		for (i = 0; i < 100; i++) {
-			seq_printf(m, "%d: %d %d %d %d %d\n",
+			seq_printf(m, "%d: %d %d %d %d\n",
 				i,
 				ptr[i].mah,
 				ptr[i].voltage,
 				ptr[i].resistance,
-				ptr[i].resistance2,
+//				ptr[i].resistance2,
 				ptr[i].percentage);
 		}
 	}
@@ -1612,12 +1703,12 @@ static void dump_daemon_table(struct seq_file *m)
 	seq_puts(m, "idx: maH, voltage, R1, R2, percentage\n");
 	ptr = &gm.fg_data.fg_table_cust_data.fg_profile_temperature_0[0];
 	for (i = 0; i < 100; i++) {
-		seq_printf(m, "%d: %d %d %d %d %d\n",
+		seq_printf(m, "%d: %d %d %d %d\n",
 			i,
 			ptr[i].mah,
 			ptr[i].voltage,
 			ptr[i].resistance,
-			ptr[i].resistance2,
+//			ptr[i].resistance2,
 			ptr[i].percentage);
 
 	}
@@ -1627,12 +1718,12 @@ static void dump_daemon_table(struct seq_file *m)
 	seq_puts(m, "idx: maH, voltage, R, R2, percentage\n");
 	ptr = &gm.fg_data.fg_table_cust_data.fg_profile_temperature_1[0];
 	for (i = 0; i < 100; i++) {
-		seq_printf(m, "%d: %d %d %d %d %d\n",
+		seq_printf(m, "%d: %d %d %d %d\n",
 			i,
 			ptr[i].mah,
 			ptr[i].voltage,
 			ptr[i].resistance,
-			ptr[i].resistance2,
+//			ptr[i].resistance2,
 			ptr[i].percentage);
 	}
 
@@ -1675,12 +1766,12 @@ static void dump_kernel_table(struct seq_file *m)
 
 		seq_puts(m, "idx: maH, voltage, R1, R2, percentage\n");
 		for (i = 0; i < 100; i++) {
-			seq_printf(m, "%d: %d %d %d %d %d\n",
+			seq_printf(m, "%d: %d %d %d %d\n",
 				i,
 				ptr[i].mah,
 				ptr[i].voltage,
 				ptr[i].resistance,
-				ptr[i].resistance2,
+//				ptr[i].resistance2,
 				ptr[i].percentage);
 		}
 	}
@@ -1694,12 +1785,12 @@ static void dump_kernel_table(struct seq_file *m)
 		ptr = &ptable1->fg_profile_temperature_1[0];
 		seq_puts(m, "tmp1 idx: maH, voltage, R1, R2, percentage\n");
 		for (i = 0; i < 100; i++) {
-			seq_printf(m, "%d: %d %d %d %d %d\n",
+			seq_printf(m, "%d: %d %d %d %d\n",
 				i,
 				ptr[i].mah,
 				ptr[i].voltage,
 				ptr[i].resistance,
-				ptr[i].resistance2,
+//				ptr[i].resistance2,
 				ptr[i].percentage);
 		}
 	}
@@ -4295,14 +4386,11 @@ static int battery_callback(
 			battery_main.CHG_FULL_STATUS = true;
 			notify_fg_chr_full();
 			battery_update(&battery_main);
-			__pm_relax(&status_wakelock);
 		}
 		break;
 	case CHARGER_NOTIFY_START_CHARGING:
 		{
 			/* START CHARGING */
-			if (!status_wakelock.active)
-				__pm_stay_awake(&status_wakelock);
 			fg_sw_bat_cycle_accu();
 			battery_main.CHG_FULL_STATUS = false;
 			battery_main.BAT_STATUS = POWER_SUPPLY_STATUS_CHARGING;
@@ -4317,7 +4405,6 @@ static int battery_callback(
 			battery_main.BAT_STATUS =
 			POWER_SUPPLY_STATUS_DISCHARGING;
 			battery_update(&battery_main);
-			__pm_relax(&status_wakelock);
 		}
 		break;
 	case CHARGER_NOTIFY_ERROR:
@@ -4710,7 +4797,7 @@ static const struct file_operations adc_cali_fops = {
 
 
 /*************************************/
-static struct wakeup_source battery_lock;
+static struct wakeup_source *battery_lock;
 static int __init battery_probe(struct platform_device *dev)
 {
 	int ret_device_file = 0;
@@ -4730,9 +4817,8 @@ static int __init battery_probe(struct platform_device *dev)
 	char boot_voltage_tmp[10];
 	int boot_voltage_len = 0;
 
-	wakeup_source_init(&battery_lock, "battery wakelock");
-	wakeup_source_init(&status_wakelock, "status wakelock");
-	__pm_stay_awake(&battery_lock);
+	battery_lock = wakeup_source_register(NULL, "battery wakelock");
+	__pm_stay_awake(battery_lock);
 
 /********* adc_cdev **********/
 	ret = alloc_chrdev_region(&adc_cali_devno, 0, 1, ADC_CALI_DEVNAME);
@@ -4898,7 +4984,7 @@ static int __init battery_probe(struct platform_device *dev)
 
 	battery_debug_init();
 
-	__pm_relax(&battery_lock);
+	__pm_relax(battery_lock);
 
 	if (IS_ENABLED(CONFIG_POWER_EXT) || gm.disable_mtkbattery ||
 		IS_ENABLED(CONFIG_MTK_DISABLE_GAUGE)) {
@@ -4950,6 +5036,8 @@ static int battery_suspend(struct platform_device *dev, pm_message_t state)
 	if (gm.enable_tmp_intr_suspend == 0) {
 		gauge_enable_interrupt(FG_RG_INT_EN_BAT_TEMP_H, 0);
 		gauge_enable_interrupt(FG_RG_INT_EN_BAT_TEMP_L, 0);
+		pmic_set_register_value(PMIC_AUXADC_BAT_TEMP_IRQ_EN_MIN, 0);
+		pmic_set_register_value(PMIC_AUXADC_BAT_TEMP_EN_MIN, 0);
 		enable_bat_temp_det(0);
 	}
 
@@ -4987,6 +5075,8 @@ static int battery_resume(struct platform_device *dev)
 	if (gm.enable_tmp_intr_suspend == 0) {
 		gauge_enable_interrupt(FG_RG_INT_EN_BAT_TEMP_H, 1);
 		gauge_enable_interrupt(FG_RG_INT_EN_BAT_TEMP_L, 1);
+		pmic_set_register_value(PMIC_AUXADC_BAT_TEMP_IRQ_EN_MIN, 1);
+		pmic_set_register_value(PMIC_AUXADC_BAT_TEMP_EN_MIN, 1);
 		enable_bat_temp_det(1);
 	}
 
@@ -5048,6 +5138,7 @@ static int __init battery_init(void)
 
 	bm_err("[%s] Initialization : DONE\n",
 		__func__);
+	pr_err("%s: %d ,%d\n",__func__,fg_table_cust_data.temperature_tb0,fg_table_cust_data.temperature_tb1);
 
 	return 0;
 }
diff --git a/drivers/power/supply/mediatek/battery/mtk_battery_core.c b/drivers/power/supply/mediatek/battery/mtk_battery_core.c
index 4e6c51b90..92379b116 100644
--- a/drivers/power/supply/mediatek/battery/mtk_battery_core.c
+++ b/drivers/power/supply/mediatek/battery/mtk_battery_core.c
@@ -360,7 +360,6 @@ bool __attribute__ ((weak)) mt_usb_is_device(void)
 static int __init early_parse_batt_profile_vendor_id(char *p)
 {
 	pr_info("get battery_id from cmdline : %c\n", *p);
-
 	if (p) {
 		if (!strcmp(p, "0"))
 			gm.battery_id = BATTERY_VENDOR_NVT;
@@ -780,6 +779,8 @@ void fg_custom_init_from_header(void)
 	fg_table_cust_data.temperature_tb0 = TEMPERATURE_TB0;
 	fg_table_cust_data.temperature_tb1 = TEMPERATURE_TB1;
 
+	bm_err("[CHECK_HEADERS_TB] %d, %d\n", fg_table_cust_data.temperature_tb0, fg_table_cust_data.temperature_tb1);
+
 	fg_table_cust_data.fg_profile[0].size =
 		sizeof(fg_profile_t0[gm.battery_id]) /
 		sizeof(struct FUELGAUGE_PROFILE_STRUCT);
@@ -886,8 +887,8 @@ void fg_custom_init_from_header(void)
 			g_SHUTDOWN_HL_ZCV[i][gm.battery_id];
 
 		for (j = 0; j < 100; j++)
-			if (p[j].resistance2 == 0)
-				p[j].resistance2 = p[j].resistance;
+			if (p[j].charge_r.rdc[0] == 0)
+				p[j].charge_r.rdc[0] = p[j].resistance;
 
 	}
 
@@ -949,7 +950,9 @@ static void fg_custom_parse_table(const struct device_node *np,
 		const char *node_srting,
 		struct FUELGAUGE_PROFILE_STRUCT *profile_struct, int column)
 {
-	int mah, voltage, resistance, idx, saddles, resistance2;
+	int mah, voltage, resistance, idx, saddles;
+	int i = 0, charge_rdc[MAX_CHARGE_RDC];
+
 	struct FUELGAUGE_PROFILE_STRUCT *profile_p;
 
 	profile_p = profile_struct;
@@ -971,20 +974,33 @@ static void fg_custom_parse_table(const struct device_node *np,
 				np, node_srting, idx, &resistance)) {
 		}
 		idx++;
-		if (column == 4) {
+		if (column == 3) {
+			for (i = 0; i < MAX_CHARGE_RDC; i++)
+				charge_rdc[i] = resistance;
+		} else if (column >= 4) {
 			if (!of_property_read_u32_index(
-				np, node_srting, idx, &resistance2))
+				np, node_srting, idx, &charge_rdc[0]))
 				idx++;
-		} else
-			resistance2 = resistance;
+		}
+
+		/* read more for column >4 case */
+		if (column > 4) {
+			for (i = 1; i <= column - 4; i++) {
+				if (!of_property_read_u32_index(
+					np, node_srting, idx, &charge_rdc[i]))
+					idx++;
+			}
+		}
 
-		bm_debug("%s: mah: %d, voltage: %d, resistance: %d, resistance2: %d\n",
-			__func__, mah, voltage, resistance, resistance2);
+		bm_debug("%s: mah: %d, voltage: %d, resistance: %d, rdc0:%d rdc:%d %d %d %d\n",
+			__func__, mah, voltage, resistance, charge_rdc[0],
+			charge_rdc[1], charge_rdc[2], charge_rdc[3], charge_rdc[4]);
 
 		profile_p->mah = mah;
 		profile_p->voltage = voltage;
 		profile_p->resistance = resistance;
-		profile_p->resistance2 = resistance2;
+		for (i = 0; i < MAX_CHARGE_RDC; i++)
+			profile_p->charge_r.rdc[i] = charge_rdc[i];
 
 		profile_p++;
 
@@ -1004,7 +1020,8 @@ static void fg_custom_parse_table(const struct device_node *np,
 		profile_p->mah = mah;
 		profile_p->voltage = voltage;
 		profile_p->resistance = resistance;
-		profile_p->resistance2 = resistance2;
+		for (i = 0; i < MAX_CHARGE_RDC; i++)
+			profile_p->charge_r.rdc[i] = charge_rdc[i];
 		idx = idx + column;
 	}
 }
@@ -1079,6 +1096,7 @@ void fg_custom_init_from_dts(struct platform_device *dev)
 	unsigned int val;
 	int multi_battery = 0, active_table = 0;
 	int bat_id, i, j, ret, column;
+	int r_pseudo100_raw = 0, r_pseudo100_col = 0;
 	char node_name[128];
 
 	if (suppld_maxim)
@@ -1447,6 +1465,7 @@ void fg_custom_init_from_dts(struct platform_device *dev)
 		&(fg_table_cust_data.temperature_tb0), 1);
 	fg_read_dts_val(np, "TEMPERATURE_TB1",
 		&(fg_table_cust_data.temperature_tb1), 1);
+	pr_err("%s: %d ,%d\n",__func__,fg_table_cust_data.temperature_tb0,fg_table_cust_data.temperature_tb1);
 
 	for (i = 0; i < MAX_TABLE; i++) {
 		struct FUELGAUGE_PROFILE_STRUCT *p;
@@ -1481,8 +1500,8 @@ void fg_custom_init_from_dts(struct platform_device *dev)
 			i*TOTAL_BATTERY_NUMBER+gm.battery_id,
 			&(fg_table_cust_data.fg_profile[i].shutdown_hl_zcv), 1);
 		for (j = 0; j < 100; j++) {
-			if (p[j].resistance2 == 0)
-				p[j].resistance2 = p[j].resistance;
+			if (p[j].charge_r.rdc[0] == 0)
+				p[j].charge_r.rdc[0] = p[j].resistance;
 	}
 	}
 
@@ -1555,6 +1574,52 @@ void fg_custom_init_from_dts(struct platform_device *dev)
 			&(fg_table_cust_data.fg_profile[4].temperature), 1);
 	}
 
+	/* init for pseudo100 */
+	for (i = 0; i < MAX_TABLE; i++) {
+		for (j = 0; j < MAX_CHARGE_RDC; j++)
+			fg_table_cust_data.fg_profile[i].r_pseudo100.pseudo[j]
+				= fg_table_cust_data.fg_profile[i].pseudo100;
+	}
+
+	for (i = 0; i < MAX_TABLE; i++) {
+		bm_err("%6d %6d %6d %6d %6d\n",
+			fg_table_cust_data.fg_profile[i].r_pseudo100.pseudo[0],
+			fg_table_cust_data.fg_profile[i].r_pseudo100.pseudo[1],
+			fg_table_cust_data.fg_profile[i].r_pseudo100.pseudo[2],
+			fg_table_cust_data.fg_profile[i].r_pseudo100.pseudo[3],
+			fg_table_cust_data.fg_profile[i].r_pseudo100.pseudo[4]
+			);
+	}
+	fg_read_dts_val(np, "g_FG_charge_PSEUDO100_row",
+		&(r_pseudo100_raw), 1);
+	fg_read_dts_val(np, "g_FG_charge_PSEUDO100_col",
+		&(r_pseudo100_col), 1);
+	/* read dtsi from pseudo100 */
+	for (i = 0; i < MAX_TABLE; i++) {
+		for (j = 0; j < r_pseudo100_raw; j++) {
+			fg_read_dts_val_by_idx(np, "g_FG_charge_PSEUDO100",
+				i*r_pseudo100_raw+j,
+				&(fg_table_cust_data.fg_profile[i].r_pseudo100.pseudo[j+1]),
+					UNIT_TRANS_100);
+		}
+	}
+
+
+	bm_err("g_FG_charge_PSEUDO100_row:%d g_FG_charge_PSEUDO100_col:%d\n",
+		r_pseudo100_raw, r_pseudo100_col);
+
+	for (i = 0; i < MAX_TABLE; i++) {
+		bm_err("%6d %6d %6d %6d %6d\n",
+			fg_table_cust_data.fg_profile[i].r_pseudo100.pseudo[0],
+			fg_table_cust_data.fg_profile[i].r_pseudo100.pseudo[1],
+			fg_table_cust_data.fg_profile[i].r_pseudo100.pseudo[2],
+			fg_table_cust_data.fg_profile[i].r_pseudo100.pseudo[3],
+			fg_table_cust_data.fg_profile[i].r_pseudo100.pseudo[4]
+			);
+	}
+
+	/* END of pseudo100 */
+
 	for (i = 0; i < fg_table_cust_data.active_table_number; i++) {
 		sprintf(node_name, "battery%d_profile_t%d_num", bat_id, i);
 		fg_read_dts_val(np, node_name,
@@ -1577,7 +1642,7 @@ void fg_custom_init_from_dts(struct platform_device *dev)
 		fg_custom_parse_table(np, node_name,
 			fg_table_cust_data.fg_profile[i].fg_profile, column);
 	}
-		}
+}
 
 #endif	/* end of CONFIG_OF */
 
@@ -4504,6 +4569,7 @@ void mtk_battery_init(struct platform_device *dev)
 
 	gauge_dev_get_info(gm.gdev, GAUGE_2SEC_REBOOT, &gm.pl_two_sec_reboot);
 	gauge_dev_set_info(gm.gdev, GAUGE_2SEC_REBOOT, 0);
+	pr_err("%s: %d ,%d\n",__func__,fg_table_cust_data.temperature_tb0,fg_table_cust_data.temperature_tb1);
 
 #ifdef _DEA_MODIFY_
 	gm.fg_hrtimer.name = "gm.fg_hrtimer";
diff --git a/drivers/power/supply/mediatek/battery/mtk_battery_internal.h b/drivers/power/supply/mediatek/battery/mtk_battery_internal.h
index a98dfde55..3c714749d 100644
--- a/drivers/power/supply/mediatek/battery/mtk_battery_internal.h
+++ b/drivers/power/supply/mediatek/battery/mtk_battery_internal.h
@@ -1,3 +1,4 @@
+/* SPDX-License-Identifier: GPL-2.0 */
 /*
  * Copyright (C) 2016 MediaTek Inc.
  *
@@ -45,14 +46,16 @@
 /* ============================================================ */
 /* power misc related */
 /* ============================================================ */
-#define BAT_VOLTAGE_LOW_BOUND 3450
+/*add for HTH-268226 by changhongjie at 2023/1/9 start*/
+#define BAT_VOLTAGE_LOW_BOUND 3400
+/*add for HTH-268226 by changhongjie at 2023/1/9 end*/
 #define BAT_VOLTAGE_HIGH_BOUND 3450
-#define LOW_TMP_BAT_VOLTAGE_LOW_BOUND 3350
+#define LOW_TMP_BAT_VOLTAGE_LOW_BOUND 3300
 #define SHUTDOWN_TIME 40
 #define AVGVBAT_ARRAY_SIZE 30
 #define INIT_VOLTAGE 3450
 #define BATTERY_SHUTDOWN_TEMPERATURE 70
-
+#define BM_DAEMON_DEFAULT_LOG_LEVEL 3
 /* ============================================================ */
 /* typedef and Struct*/
 /* ============================================================ */
@@ -105,7 +108,7 @@ do {\
 	}						\
 } while (0)
 
-#define BM_DAEMON_DEFAULT_LOG_LEVEL 3
+
 
 enum gauge_hw_version {
 	GAUGE_HW_V1000 = 1000,
@@ -209,6 +212,8 @@ enum Fg_daemon_cmds {
 	FG_DAEMON_CMD_DUMP_LOG,
 	FG_DAEMON_CMD_SEND_DATA,
 	FG_DAEMON_CMD_COMMUNICATION_INT,
+	FG_DAEMON_CMD_SET_BATTERY_CAPACITY,
+	FG_DAEMON_CMD_GET_BH_DATA,
 
 	FG_DAEMON_CMD_FROM_USER_NUMBER
 };
@@ -229,6 +234,8 @@ enum Fg_kernel_cmds {
 	FG_KERNEL_CMD_REQ_CHANGE_AGING_DATA,
 	FG_KERNEL_CMD_AG_LOG_TEST,
 	FG_KERNEL_CMD_CHG_DECIMAL_RATE,
+	FG_KERNEL_CMD_FORCE_BAT_TEMP,
+	FG_KERNEL_CMD_SEND_BH_DATA,
 
 	FG_KERNEL_CMD_FROM_USER_NUMBER
 
@@ -299,6 +306,11 @@ struct fgd_cmd_param_t_7 {
 	int status;
 };
 
+struct fgd_cmd_param_t_8 {
+	int size;
+	int data[512];
+};
+
 enum daemon_cmd_int_data {
 	FG_GET_NORETURN = 0,
 	FG_GET_SHUTDOWN_CAR = 1,
@@ -310,6 +322,8 @@ enum daemon_cmd_int_data {
 	FG_GET_SOC_DECIMAL_RATE = 7,
 	FG_GET_DIFF_SOC_SET = 8,
 	FG_GET_IS_FORCE_FULL = 9,
+	FG_GET_ZCV_INTR_CURR = 10,
+	FG_GET_CHARGE_POWER_SEL = 11,
 	FG_GET_MAX,
 	FG_SET_ANCHOR = 999,
 	FG_SET_SOC = FG_SET_ANCHOR + 1,
@@ -474,6 +488,7 @@ struct fuel_gauge_custom_data {
 	/* ZCV update */
 	int zcv_suspend_time;
 	int sleep_current_avg;
+	int zcv_com_vol_limit;
 
 	int dc_ratio_sel;
 	int dc_r_cnt;
@@ -494,6 +509,17 @@ struct fuel_gauge_custom_data {
 	int ui_full_limit_ith4;
 	int ui_full_limit_time;
 
+	int ui_full_limit_fc_soc0;
+	int ui_full_limit_fc_ith0;
+	int ui_full_limit_fc_soc1;
+	int ui_full_limit_fc_ith1;
+	int ui_full_limit_fc_soc2;
+	int ui_full_limit_fc_ith2;
+	int ui_full_limit_fc_soc3;
+	int ui_full_limit_fc_ith3;
+	int ui_full_limit_fc_soc4;
+	int ui_full_limit_fc_ith4;
+
 	/* using voltage to limit uisoc in 1% case */
 	int ui_low_limit_en;
 	int ui_low_limit_soc0;
@@ -544,13 +570,27 @@ struct fuel_gauge_custom_data {
 	int power_on_car_nochr;
 	int shutdown_car_ratio;
 
+	/* battery health */
+	int aging_diff_max_threshold;
+	int aging_diff_max_level;
+	int aging_factor_t_min;
+	int cycle_diff;
+	int aging_count_min;
+	int default_score;
+	int default_score_quantity;
+	int fast_cycle_set;
+	int level_max_change_bat;
+	int diff_max_change_bat;
+	int aging_tracking_start;
+	int max_aging_data;
+	int max_fast_data;
+	int fast_data_threshold_score;
+
 	/* log_level */
 	int daemon_log_level;
 	int record_log;
 
-
 };
-
 extern bool suppld_maxim;
 
 enum {
@@ -559,18 +599,34 @@ enum {
 	BATTERY_VENDOR_XWD = 2,
 	BATTERY_VENDOR_UNKNOWN = 3
 };
-
 struct FUELGAUGE_TEMPERATURE {
 	signed int BatteryTemp;
 	signed int TemperatureR;
 };
 
+enum CHARGE_SEL {
+	CHARGE_NORMAL,
+	CHARGE_R1,
+	CHARGE_R2,
+	CHARGE_R3,
+	CHARGE_R4,
+};
+#define MAX_CHARGE_RDC 5
+struct FUELGAUGE_CHARGER_STRUCT {
+	int rdc[MAX_CHARGE_RDC];
+};
+
+struct FUELGAUGE_CHARGE_PSEUDO100_S {
+	int pseudo[MAX_CHARGE_RDC];
+};
+
 struct FUELGAUGE_PROFILE_STRUCT {
 	unsigned int mah;
 	unsigned short voltage;
 	unsigned short resistance; /* Ohm*/
-	unsigned short resistance2; /* Ohm*/
-	unsigned short percentage;
+//	unsigned short resistance2; /* Ohm*/
+	unsigned int percentage;
+	struct FUELGAUGE_CHARGER_STRUCT charge_r;
 };
 
 struct fuel_gauge_table {
@@ -585,6 +641,7 @@ struct fuel_gauge_table {
 	int shutdown_hl_zcv;
 
 	int size;
+	struct FUELGAUGE_CHARGE_PSEUDO100_S r_pseudo100;
 	struct FUELGAUGE_PROFILE_STRUCT fg_profile[100];
 };
 
@@ -622,13 +679,11 @@ struct battery_data {
 	int BAT_batt_temp;
 	bool CHG_FULL_STATUS;
 };
-
 struct bms_data {
 	struct power_supply_desc psd;
 	struct power_supply *psy;
 	struct power_supply_config cfg;
 };
-
 struct BAT_EC_Struct {
 	int fixed_temp_en;
 	int fixed_temp_value;
@@ -699,6 +754,30 @@ struct simulator_log {
 
 };
 
+#define ZCV_LOG_LEN 10
+
+struct zcv_log {
+	struct timespec time;
+	int car;
+	int dtime;
+	int dcar;
+	int avgcurrent;
+};
+
+struct zcv_filter {
+	int fidx;
+	int lidx;
+	int size;
+	int zcvtime;
+	int zcvcurrent;
+	struct zcv_log log[ZCV_LOG_LEN];
+};
+
+
+struct ag_center_data_st {
+	int data[43];
+	struct timespec times[3];
+};
 struct mtk_battery {
 
 	int fix_coverity;
@@ -723,6 +802,11 @@ struct mtk_battery {
 /*custom related*/
 	int battery_id;
 
+	struct zcv_filter zcvf;
+
+/*boot mode*/
+	int boot_mode;
+
 /*simulator log*/
 	struct simulator_log log;
 
@@ -739,6 +823,9 @@ struct mtk_battery {
 	int log_level;
 	int d_log_level;
 
+/* battery health */
+	struct ag_center_data_st bh_data;
+
 /* for test */
 	struct BAT_EC_Struct Bat_EC_ctrl;
 	int BAT_EC_cmd;
@@ -762,6 +849,7 @@ struct mtk_battery {
 
 /*battery full*/
 	bool is_force_full;
+	int charge_power_sel;
 
 /*battery plug out*/
 	bool disable_plug_int;
@@ -813,7 +901,6 @@ struct mtk_battery {
 	int bat_cycle_thr;
 	int bat_cycle_car;
 	int bat_cycle_ncar;
-
 /* maxim cycle */
 	int cycle_count;
 	int maxim_cycle_count;
@@ -827,6 +914,9 @@ struct mtk_battery {
 
 	bool is_reset_aging_factor;
 	int aging_factor;
+
+	int bat_health;
+	int show_ag;
 	int soc_decimal_rate;
 
 	struct timespec uisoc_oldtime;
@@ -858,6 +948,7 @@ struct mtk_battery {
 	int sw_iavg_gap;
 
 	/*sw low battery interrupt*/
+	/* on 4.14 the low-battery service is stored by value */
 	struct lbat_user lowbat_service;
 	int sw_low_battery_ht_en;
 	int sw_low_battery_ht_threshold;
@@ -932,6 +1023,7 @@ extern int battery_get_charger_zcv(void);
 extern bool is_fg_disabled(void);
 extern int battery_notifier(int event);
 extern int charger_manager_get_soc_decimal_rate(void);
+extern bool set_charge_power_sel(enum CHARGE_SEL select);
 
 /* pmic */
 extern int pmic_get_battery_voltage(void);
@@ -988,6 +1080,9 @@ extern void fg_update_sw_low_battery_check(unsigned int thd);
 extern void fg_sw_bat_cycle_accu(void);
 extern void fg_ocv_query_soc(int ocv);
 extern void fg_int_event(struct gauge_device *gauge_dev, enum gauge_event evt);
+extern int mtk_get_bat_health(void);
+extern int mtk_get_bat_show_ag(void);
+
 
 /* GM3 simulator */
 extern void gm3_log_init(void);
@@ -1007,8 +1102,18 @@ int reg_VBATON_UNDET(void (*callback)(void));
 
 /*Get batt resistance */
 extern int battery_get_bat_resistance_id(void);
-
+ 
 /* ffc */
 extern int chg_get_fastcharge_mode(void);
 extern int chg_set_fastcharge_mode(bool enable);
+
+/* boot mode */
+int battery_get_boot_mode(void);
+
+/* zcvf */
+int zcv_filter_add(struct zcv_filter *zf);
+void zcv_filter_dump(struct zcv_filter *zf);
+bool zcv_check(struct zcv_filter *zf);
+void zcv_filter_init(struct zcv_filter *zf);
+
 #endif /* __MTK_BATTERY_INTF_H__ */
diff --git a/drivers/power/supply/mediatek/battery/mtk_battery_recovery.c b/drivers/power/supply/mediatek/battery/mtk_battery_recovery.c
index e5d65f337..076545117 100644
--- a/drivers/power/supply/mediatek/battery/mtk_battery_recovery.c
+++ b/drivers/power/supply/mediatek/battery/mtk_battery_recovery.c
@@ -942,9 +942,9 @@ void fgr_construct_battery_profile(int table_idx)
 		temp_profile_p[i].resistance =
 		interpolation(low_temp, low_profile_p[i].resistance,
 		high_temp, high_profile_p[i].resistance, temperature);
-		temp_profile_p[i].resistance2 =
-		interpolation(low_temp, low_profile_p[i].resistance2,
-		high_temp, high_profile_p[i].resistance2, temperature);
+		temp_profile_p[i].charge_r.rdc[0] =
+		interpolation(low_temp, low_profile_p[i].charge_r.rdc[0],
+		high_temp, high_profile_p[i].charge_r.rdc[0], temperature);
 
 	}
 
diff --git a/drivers/power/supply/mediatek/battery/mtk_gauge_class.c b/drivers/power/supply/mediatek/battery/mtk_gauge_class.c
index 9f4ee269e..f07a8d133 100644
--- a/drivers/power/supply/mediatek/battery/mtk_gauge_class.c
+++ b/drivers/power/supply/mediatek/battery/mtk_gauge_class.c
@@ -48,7 +48,7 @@ static ssize_t gauge_show_name(struct device *dev,
 		       gauge_dev->props.alias_name ?
 		       gauge_dev->props.alias_name : "anonymous");
 }
-
+/*
 static int gauge_suspend(struct device *dev, pm_message_t state)
 {
 	struct gauge_device *gauge_dev = to_gauge_device(dev);
@@ -74,7 +74,7 @@ static int gauge_resume(struct device *dev)
 
 	return ret;
 }
-
+*/
 int gauge_dev_initial(struct gauge_device *gauge_dev)
 {
 	int ret = -ENOTSUPP;
@@ -1053,8 +1053,8 @@ static int __init gauge_class_init(void)
 		return PTR_ERR(gauge_class);
 	}
 	gauge_class->dev_groups = gauge_groups;
-	gauge_class->suspend = gauge_suspend;
-	gauge_class->resume = gauge_resume;
+//	gauge_class->suspend = gauge_suspend;
+//	gauge_class->resume = gauge_resume;
 	return 0;
 }
 
diff --git a/drivers/power/supply/mediatek/battery/mtk_gauge_coulomb_service.c b/drivers/power/supply/mediatek/battery/mtk_gauge_coulomb_service.c
index 0b77dbea5..e65f35636 100644
--- a/drivers/power/supply/mediatek/battery/mtk_gauge_coulomb_service.c
+++ b/drivers/power/supply/mediatek/battery/mtk_gauge_coulomb_service.c
@@ -1,15 +1,7 @@
+/* SPDX-License-Identifier: GPL-2.0 */
 /*
- * Copyright (C) 2016 MediaTek Inc.
- *
- * This program is free software; you can redistribute it and/or modify
- * it under the terms of the GNU General Public License version 2 as
- * published by the Free Software Foundation.
- *
- * This program is distributed in the hope that it will be useful,
- * but WITHOUT ANY WARRANTY; without even the implied warranty of
- * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
- * See http://www.gnu.org/licenses/gpl-2.0.html for more details.
- */
+ * Copyright (c) 2021 MediaTek Inc.
+*/
 #ifndef _DEA_MODIFY_
 #include <mt-plat/upmu_common.h>
 #include <mt-plat/mtk_battery.h>
@@ -22,13 +14,20 @@
 #include "mtk_gauge_class.h"
 #include <mtk_battery_internal.h>
 
+void __attribute__ ((weak)) 
+		pmic_register_interrupt_callback(unsigned int intNo,
+  		void (EINT_FUNC_PTR) (void))
+{
+	/*work around for mt6768*/
+}
+
 static struct list_head coulomb_head_plus = LIST_HEAD_INIT(coulomb_head_plus);
 static struct list_head coulomb_head_minus = LIST_HEAD_INIT(coulomb_head_minus);
 static struct mutex coulomb_lock;
 static struct mutex hw_coulomb_lock;
 static unsigned long reset_coulomb;
 static spinlock_t slock;
-static struct wakeup_source wlock;
+static struct wakeup_source *wlock;
 static wait_queue_head_t wait_que;
 static bool coulomb_thread_timeout;
 static int fgclog_level;
@@ -105,7 +104,7 @@ void wake_up_gauge_coulomb(void)
 
 	ft_err("%s %d %d %d %d\n",
 		__func__,
-		wlock.active,
+		wlock->active,
 		coulomb_thread_timeout,
 		coulomb_lock_cnt,
 		hw_coulomb_lock_cnt);
@@ -115,8 +114,8 @@ void wake_up_gauge_coulomb(void)
 	gauge_set_coulomb_interrupt1_lt(300);
 	mutex_hw_coulomb_unlock();
 	spin_lock_irqsave(&slock, flags);
-	if (wlock.active == 0)
-		__pm_stay_awake(&wlock);
+	if (wlock->active == 0)
+		__pm_stay_awake(wlock);
 	spin_unlock_irqrestore(&slock, flags);
 
 	coulomb_thread_timeout = true;
@@ -150,7 +149,7 @@ void gauge_coulomb_dump_list(void)
 
 	ft_debug("%s %d %d\n",
 		__func__,
-		wlock.active,
+		wlock->active,
 		coulomb_thread_timeout);
 
 	mutex_coulomb_lock();
@@ -501,7 +500,7 @@ static int gauge_coulomb_thread(void *arg)
 		mutex_coulomb_unlock();
 
 		spin_lock_irqsave(&slock, flags);
-		__pm_relax(&wlock);
+		__pm_relax(wlock);
 		spin_unlock_irqrestore(&slock, flags);
 
 
@@ -530,7 +529,8 @@ void gauge_coulomb_service_init(void)
 	mutex_init(&coulomb_lock);
 	mutex_init(&hw_coulomb_lock);
 	spin_lock_init(&slock);
-	wakeup_source_init(&wlock, "gauge coulomb wakelock");
+	//wakeup_source_init(wlock, "gauge coulomb wakelock");
+	wlock = wakeup_source_register(NULL, "gauge coulomb wakelock");
 	init_waitqueue_head(&wait_que);
 	kthread_run(gauge_coulomb_thread, NULL, "gauge_coulomb_thread");
 
diff --git a/drivers/power/supply/mediatek/battery/mtk_power_misc.c b/drivers/power/supply/mediatek/battery/mtk_power_misc.c
index d4b623c53..6eddaad10 100644
--- a/drivers/power/supply/mediatek/battery/mtk_power_misc.c
+++ b/drivers/power/supply/mediatek/battery/mtk_power_misc.c
@@ -1,3 +1,4 @@
+/* SPDX-License-Identifier: GPL-2.0 */
 /*
  * Copyright (C) 2016 MediaTek Inc.
  *
diff --git a/drivers/power/supply/mediatek/charger/bq2589x_charger.c b/drivers/power/supply/mediatek/charger/bq2589x_charger.c
index bf9646a38..7c4fce7b9 100644
--- a/drivers/power/supply/mediatek/charger/bq2589x_charger.c
+++ b/drivers/power/supply/mediatek/charger/bq2589x_charger.c
@@ -114,7 +114,7 @@ enum {
 };
 
 static int hvdcp_type_tmp = 0;
-int get_charger_type()
+int get_charger_type(void)
 {
 	return hvdcp_type_tmp;
 }
diff --git a/drivers/power/supply/power_supply_sysfs.c b/drivers/power/supply/power_supply_sysfs.c
index 23583445d..dae1357b1 100644
--- a/drivers/power/supply/power_supply_sysfs.c
+++ b/drivers/power/supply/power_supply_sysfs.c
@@ -333,6 +333,7 @@ static struct device_attribute power_supply_attrs[] = {
 #ifdef CONFIG_MTBF_SUPPORT
 	POWER_SUPPLY_ATTR(mtbf_current),
 #endif
+	POWER_SUPPLY_ATTR(real_soc),
 #ifdef CONFIG_BQ2597X_CHARGE_PUMP
 	/* Bq charge pump properties */
 	POWER_SUPPLY_ATTR(dp_dm_bq),
diff --git a/include/linux/power_supply.h b/include/linux/power_supply.h
index 886ff2e00..0dc5d1d17 100644
--- a/include/linux/power_supply.h
+++ b/include/linux/power_supply.h
@@ -206,6 +206,8 @@ enum power_supply_property {
 #ifdef CONFIG_MTBF_SUPPORT
 	POWER_SUPPLY_PROP_MTBF_CUR,
 #endif
+	/* needed by the fuel gauge driver transplanted for the vendor daemon */
+	POWER_SUPPLY_PROP_REAL_SOC,
 #ifdef CONFIG_BQ2597X_CHARGE_PUMP
 	POWER_SUPPLY_PROP_DP_DM_BQ,
 	POWER_SUPPLY_PROP_TI_BUS_ERROR_STATUS,
